File: 8-linked-list/code.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Struktur Node untuk Linked List
typedef struct Node {
    int data;
    struct Node* next;
} Node;

// Fungsi untuk membuat node baru
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        printf("Memory allocation failed\n");
        exit(1);
    }
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Fungsi untuk menambah node di awal linked list
Node* insertAtBeginning(Node* head, int data) {
    Node* newNode = createNode(data);
    newNode->next = head;
    return newNode;
}

// Fungsi untuk mencari node dalam linked list
Node* search(Node* head, int key) {
    Node* current = head;
    while (current != NULL) {
        if (current->data == key) {
            return current;
        }
        current = current->next;
    }
    return NULL;
}

// Fungsi untuk menghapus linked list
void freeList(Node* head) {
    Node* temp;
    while (head != NULL) {
        temp = head;
        head = head->next;
        free(temp);
    }
}

int main() {
    clock_t start, end;
    double cpu_time_used;
    
    // Inisialisasi linked list
    Node* head = NULL;
    
    start = clock();
    
    // Operasi pada linked list
    // Menambahkan node (misal 1000 node)
    for (int i = 0; i < 1000; i++) {
        head = insertAtBeginning(head, i);
    }
    
    // Mencari beberapa nilai
    for (int i = 0; i < 100; i++) {
        int key = rand() % 1000;
        Node* result = search(head, key);
        if (result != NULL) {
            printf("Found %d\n", key);
        } else {
            printf("%d not found\n", key);
        }
    }
    
    // Membersihkan memori
    freeList(head);
    
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("Waktu eksekusi: %f detik\n", cpu_time_used);
    
    return 0;
}

File: 8-linked-list/code.rs
use std::time::Instant;

// Definisi Node untuk Linked List
struct Node {
    data: i32,
    next: Option<Box<Node>>,
}

// Implementasi untuk Node
impl Node {
    // Membuat node baru
    fn new(data: i32) -> Self {
        Node {
            data,
            next: None,
        }
    }
}

// Struktur LinkedList untuk mengelola linked list
struct LinkedList {
    head: Option<Box<Node>>,
}

// Implementasi untuk LinkedList
impl LinkedList {
    // Membuat linked list baru
    fn new() -> Self {
        LinkedList {
            head: None,
        }
    }
    
    // Menambahkan node di awal
    fn insert_at_beginning(&mut self, data: i32) {
        let mut new_node = Box::new(Node::new(data));
        new_node.next = self.head.take();
        self.head = Some(new_node);
    }
    
    // Mencari node dalam linked list
    fn search(&self, key: i32) -> bool {
        let mut current = &self.head;
        while let Some(node) = current {
            if node.data == key {
                return true;
            }
            current = &node.next;
        }
        false
    }
}

fn main() {
    let start = Instant::now();
    
    // Inisialisasi linked list
    let mut list = LinkedList::new();
    
    // Menambahkan node (misal 1000 node)
    for i in 0..1000 {
        list.insert_at_beginning(i);
    }
    
    for i in 0..100 {
        let key = (i * 17) % 1000;
        let found = list.search(key);
        if found {
            println!("Found {}", key);
        } else {
            println!("{} not found", key);
        }
    }
    
    // Linked list akan dibersihkan secara otomatis ketika keluar dari scope
    
    let duration = start.elapsed();
    println!("Waktu eksekusi: {:?}", duration);
}